### ISSUE TITLE
Flutter is "unable to locate a development device" despite the fact that "flutter devices" sees it.

### ISSUE THREAD

**Issue by mkustermann** [flutter/flutter#2749]
Flutter is "unable to locate a development device" despite the fact that "flutter devices" sees it. The recommendation to run `flutter doctor` is not helpful in any way.
### Steps to Reproduce

```
myfirechat $ flutter run --checked --debug-port=16162 --start-paused --device-id 00cbeb501a994ff4
Unable to locate a development device; please run 'flutter doctor' for information about installing additional components.

myfirechat $ flutter doctor                                                                       
Flutter root: .../flutter-hackathon.

[✓] Android toolchain - develop for Android devices (installed)
  [✓] Android Studio / Android SDK - enable development for Android devices (installed)

[✓] Atom development environment - a lightweight development environment for Flutter (installed)
  [✓] Atom editor (installed)
  [✓] Flutter plugin - adds Flutter specific functionality to Atom (installed)

Flutter from git@github.com:flutter/flutter.git (on hackathon)
Framework: e1f0d8c974a58977e19ab685d2eb323d2d9becd4 

myfirechat $ flutter devices
1 connected device:

Nexus 5X • 00cbeb501a994ff4 • android-arm
```
### Flutter Version

```
myfirechat $  flutter --version
Flutter from git@github.com:flutter/flutter.git (on hackathon)
Framework: e1f0d8c974a58977e19ab685d2eb323d2d9becd4 (30 hours ago)
Engine:    5e68565d713cbb89e0d60e4c635cf7a9e4b2b855

[✓] The Android toolchain is fully installed.
[✓] The Atom development environment is fully installed.
```

**Comment by mkustermann** [comment]
After spending time debugging this in the flutter sources, it turns out this has to do with some version matching logic, which fails silently.

More specifically:
```
$ ls $ANDROID_HOME/build-tools
24.0.0-preview/

$ ls $ANDROID_HOME/platforms
android-10  android-15  android-16  android-17  android-18  android-19  android-20  android-21  android-22  android-23  android-8  android-N
```

It tries to match `24` with all of the other versions after `android-xxx` which fails.

**Comment by devoncarew** [comment]
Thanks for debugging it! I think we're able to locate devices because that only depends on finding `adb`. Later, when we run, that requires a full android sdk (or enough bits to build an android app).

**Comment by devoncarew** [comment]
Although, you should not see this: `[✓] The Android toolchain is fully installed.`

**Comment by devoncarew** [comment]
@mkustermann the fix for this has landed in master. If you could test it locally that would be great.

**Comment by mkustermann** [comment]
@devoncarew 

First nothing changed, even after a `flutter upgrade`. I suspect that this is due to the fact that flutter uses a git repository setup and it somehow remembers the branch and the `origin/hackathon` branch hasn't been updated with your fix.

If I manually reset to `origin/master` and run `flutter upgrade` again, it will take `origin/hackathon` and replay all commits from `origin/hackathon..origin/master` on top of it.

=> This is flutter git setup is a bit confusing / weired!

The good news: Now I got even the last piece of the codelab working, namely firebase. That is pretty cool :)

**Comment by devoncarew** [comment]
Cool, great to hear.

Yeah, the flutter channels are just git branches. We expect most people to work on `alpha` for now; that's updated from `master` periodically. `hackathon` we keep synced with the codelab.

**Comment by alvaroc1** [comment]
I'm running into the same issue. @devoncarew How do I get the fixed version? I am running master AFAICT. Could it be caused by a different underlying problem? @mkustermann Where did you create the symlinks?

**Comment by devoncarew** [comment]
@alvaroc1, can you open a new issue? It may be similar to this one, which looks like it was a problem with us locating all the components we needed from the android sdk. Some things that would help triage the new issue would be the output of flutter doctor, and the listing of the dirs in the android-sdk/platforms and android-sdk/build-tools. Thanks!

**Comment by ajdewit1994** [comment]
Sorry for digging in old topics, but I'm having the same problem. Flutter tells me that there are no devices to be found, yet shows them under "flutter devices". How can I fix this?

**Comment by Hixie** [comment]
@ajdewit1994 I've reopened this bug for now but if you wouldn't mind can you file a new one and include all the details you can about your setup?

**Comment by pe-pan** [comment]
I'm having the very same issue.

**Comment by eseidelGoogle** [comment]
I wonder if the --device-id based selection isn't working. I use `-d` with a partial name, so with your case i would use `-d n` to select the "Nexus 5" device.  @tvolkert would know.

**Comment by tvolkert** [comment]
Yes, `-d n` should work.  @pe-pan can you run with `--bug-report` and attach the associated zip file here?  Note that it may end up including part of the source code for your app and local file system paths and such -- if that's a concern, you'll want to scrub it or send it to me privately.

**Comment by jxmorris12** [comment]
I'm having this issue as well. I suspect issues with the Android SDK install.

**Comment by tvolkert** [comment]
@jxmorris12 did you have an `ANDROID_SDK` environment variable set?  And I'm guessing you didn't have an `ANDROID_HOME` environment variable set?

**Comment by jxmorris12** [comment]
@tvolkert I don't think that either variable was set.

**Comment by Hixie** [comment]
I'm going to close this bug because it's not actionable for us in its current state.

If you see this behaviour, please file a new bug with the following information:

* the output of `flutter run --bug-report` and `flutter doctor --bug-report`.
* information about what versions of the Android SDK you have installed.
* the values of your environment variables (especially ANDROID_SDK and ANDROID_HOME).

**Comment by akashandroid90** [comment]
I have configured value for ANDROID_HOME but i don't what value i need to set for ANDROID_SDK. Can you please help me.

**Comment by paolomoschini** [comment]
@akashandroid90 could u try to launch in a terminal the following commands:
```
$ flutter config --android-sdk /path/to/android/sdk
$ flutter config --android-studio-dir /path/to/android/studio
```
Then restart Android Studio

**Comment by kraulain** [comment]
@paolomoschini  please what is the equivalent of 
```flutter config --android-studio-dir /path/to/android/studio```
if you are using IntelliJ?

**Comment by hilbix** [comment]
Apparently IntelliJ does not set `$ANDROID_HOME`.  `flutter doctor` did not help in this respect, as the hint to set `$ANDROID_HOME` is not helpful.  Only comment @akashandroid90 on this closed bug helped.  Following command fixed the issue for me (as the Android SDK is under `$HOME/SDK` at my side):

    flutter config --android-sdk "$HOME/SDK/android-sdk-linux"

Thank you for this correct pointer.

Notes:

- Setting a global variable and restarting X11 is not helpful, as you might want to work with several Android-SDK variants in parallel, where only (typically the wrong) one can live in the environment.

- `flutter config` saves the setting into `$HOME/.flutter_settings`.  This is bad for following respect:
  - You can only select one single Android-SDK out of all installed ones, this makes it impossible to use different Android-SDKs with `flutter` in parallel
  - This setting is shared between all `flutter` variants, so the Android-SDK must fit for all Flutter-Variants you use in parallel

Think about `git`.  `git config` has 3 possible locations:

- `.git/config`
- `$HOME/.gitconfig`
- `/etc/gitconfig`

Flutter should do it similarly.  So there shall be following

- `flutter config local` which is the default, stored in `$FLUTTER_HOME/.flutter_settings` (or similar)
- `flutter config global` which is `$HOME/.flutter_settings`
- `flutter config system` which is `/etc/flutter_settings` (or similar)

But this exposes another problem:

`flutter config` states "To remove a setting, configure it to an empty string.".  There must be a way to unset a default value, such that you can unset a "global" setting in "local" context.  But there also must be some way to inherit a setting.

My recommendation about this is as follows:

- An empty string does inheritance (removes the setting)
- String `-` does inhibit inheritance and
- any other string is taken literally,
- except for strings, which are made of only `-` characters.  For those, the first `-` is removed.

TODOs:

I am using `Flutter 0.9.4 • channel beta`.  Find out if all babble above is properly addressed by the latest `flutter`.  If not, eventually:

- Open a bug, that `flutter doctor` does point to `$ANDROID_HOME` where it shall point to `flutter config` instead
- Open a bug, that `flutter config` only allows one single global config per luser and recommend to do it like `git` and explain inheritance as stated in this comment.
- Perhaps file a bug against the `flutter` plugin of IntelliJ that it should set `$ANDROID_HOME` to the one in use

Perhaps somebody else can check and do this (I currently lack the time to do this all.)

Reason why I post this here:

This bug should not have been closed with the given explanation.  And the post by @paolomoschini only gives a (good) workaround, but this still is not really a solution to the problem.

The real bug is
- that `flutter doctor` points into the wrong direction and
- `flutter config` only allows one single global setting
- and (perhaps) the `flutter` plugin of IntelliJ/AndroidStudio does not set `$ANDROID_HOME`

**Comment by yashthakkar1173** [comment]
> 
> 
> > @akashandroid90 could u try to launch in a terminal the following commands:
> > ```
> > $ flutter config --android-sdk /path/to/android/sdk
> > $ flutter config --android-studio-dir /path/to/android/studio
> > ```
> > Then restart Android Studio
> 
> I did this now android studio don't detect my device anymore

same problem!!

**Comment by adelbenhamadi** [comment]
If flutter doctor return something like this:
X Android SDK file not found: C:\Android\android-sdk\platforms\android-29\android.jar.
you have to either remove the android-29 folder if you don't need it or install the 29 sdk
